### PR TITLE
Extract build finalization to job

### PR DESCRIPTION
**Pull request by thorncp**
`BuildReportJob` should stand in place of the GitHub interactions in
`Review` in #727 with the plan of using it in the build workers
controller. Also change `Commenter` to accept a `Payload` and create its
own `PullRequest` instance.

Up next will be implementing the `BuildWorkers` controller
that will check for build completion, persist violations, and enqueue
this job.

https://trello.com/c/CdExHIVN

### CODE REVIEW

File: app/jobs/build_report_job.rb [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  end
+
+  def track_subscribed_build_completed
+    if build.repo.subscription
+      user = build.repo.subscription.user
+      analytics = Analytics.new(user)
+      analytics.track_build_completed(build.repo)
+    end
+  end
+end
```
Comment (gylaz) [inline app/jobs/build_report_job.rb]: This job is pretty hefty on logic -- thoughts on pulling this out into a model/service?

File: app/services/build_runner.rb [hunk 2]
```diff
@@ -8,14 +6,14 @@ def run
...
-        violations: violations,
+      build = repo.builds.create!(
         pull_request_number: payload.pull_request_number,
         commit_sha: payload.head_sha,
       )
-      commenter.comment_on_violations(priority_violations)
-      create_success_status
-      track_subscribed_build_completed
+
+      build.update!(violations: find_violations)
```
Comment (gylaz) [inline app/services/build_runner.rb]: Can this be combined with the `create!` call?

File: app/services/build_runner.rb [hunk 3]
```diff
@@ -27,22 +25,14 @@ def relevant_pull_request?
     pull_request.opened? || pull_request.synchronize?
   end
 
-  def violations
+  def find_violations
```
Comment (gylaz) [inline app/services/build_runner.rb]: Any particular reason for this rename? It doesn't really reflect the nature of `find`.

File: app/jobs/build_report_job.rb [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+      "pull_request" => {
+        "head" => {
+          "sha" => build.commit_sha,
+        },
+      },
+      "repository" => {
+        "full_name" => build.repo.full_github_name,
+      },
+    )
+  end
```
Comment (gylaz) [inline app/jobs/build_report_job.rb]: If we ever need to pull more information from the original payload data, we'd have to modify it in two places. How about either passing `payload` to this job, or storing the payload on the build?

File: app/services/build_runner.rb [hunk 5]
```diff
@@ -8,14 +6,14 @@ def run
...
         pull_request_number: payload.pull_request_number,
         commit_sha: payload.head_sha,
       )
-      commenter.comment_on_violations(priority_violations)
-      create_success_status
-      track_subscribed_build_completed
+
+      build.update!(violations: find_violations)
+
+      BuildReportJob.perform_later(build)
```
Comment (gylaz) [inline app/services/build_runner.rb]: `BuildRunner` already runs as a job. Is there a reason for spawning another job here?

File: app/jobs/build_report_job.rb (hunk 1 above)
Comment (teoljungberg) [inline app/jobs/build_report_job.rb]: It also follows our current style, having a job that just calls a service

File: app/services/build_runner.rb (hunk 5 above)
Comment (teoljungberg) [inline app/services/build_runner.rb]: Fixed and pushed

File: app/services/build_runner.rb (hunk 3 above)
Comment (teoljungberg) [inline app/services/build_runner.rb]: I agree, I renamed it and pushed it

File: app/services/build_runner.rb (hunk 2 above)
Comment (teoljungberg) [inline app/services/build_runner.rb]: It can, done and pushed

File: app/jobs/build_report_job.rb (hunk 1 above)
Comment (teoljungberg) [inline app/jobs/build_report_job.rb]: Done and pushed

File: app/jobs/build_report_job.rb (hunk 4 above)
Comment (teoljungberg) [inline app/jobs/build_report_job.rb]: I'm unsure what's the best way of solving this. It kinda feels like we should pass the build (that has most of this information to the `Commenter`

Comment (thorncp) [inline app/jobs/build_report_job.rb]: I like the idea of storing it on the build. When the workers are introduced, we won't have to pass the payload (or a subset of it) back and forth between Hound an Iron, since we can just pull it from the build. That should probably be a different PR though.

File: app/services/build_runner.rb (hunk 5 above)
Comment (thorncp) [inline app/services/build_runner.rb]: I think this was done with the workers in mind, where the build worker controller would kick off the job when it gets a payload from Iron. I like it better as a service, and we can another job for the build worker case.

File: app/services/build_runner.rb (hunk 3 above)
Comment (thorncp) [inline app/services/build_runner.rb]: I'm having a hard to remembering what our thought process was here. I think we were trying to indicate the severity of this method call, in that it goes through many levels of abstraction and does quite a lot of work. I don't think `find_violations` captures that, but I do think `violations` sort of trivializes what's happening.

File: app/services/commenter.rb [hunk 6]
```diff
@@ -15,4 +15,8 @@ def comment_on_violations(violations)
   def commenting_policy
     @commenting_policy ||= CommentingPolicy.new(pull_request)
   end
+
+  def token
+    ENV["HOUND_GITHUB_TOKEN"]
+  end
```
Comment (gylaz) [inline app/services/commenter.rb]: We don't need this anymore, right?

File: spec/services/build_report_spec.rb [hunk 7]
```diff
@@ -0,0 +1,111 @@
+require "rails_helper"
+
+describe BuildReport, ".run" do
```
Comment (gylaz) [inline spec/services/build_report_spec.rb]: We have a guideline against doing this:

> Use a single, top-level describe ClassName block.

File: spec/services/commenter_spec.rb [hunk 8]
```diff
@@ -9,6 +9,7 @@
         it "comments on the violation" do
           pull_request = double(:pull_request, comment_on_violation: nil)
           violation = double(:violation)
+          payload = Payload.new({})
```
Comment (gylaz) [inline spec/services/commenter_spec.rb]: Left over?

**Comment by gylaz** [comment]
This is looking great! Couple of small things, then good to merge.

**Comment by teoljungberg** [comment]
I rebased against master to get the changes made in #774 and rewrote the commit message. Thoughts @gylaz and @thorncp ?

File: app/services/build_report.rb [hunk 9]
```diff
@@ -0,0 +1,55 @@
+require "attr_extras"
+
+class BuildReport
+  MAX_COMMENTS = ENV.fetch("MAX_COMMENTS").to_i
+
+  static_facade :run, :pull_request, :build
```
Comment (gylaz) [inline app/services/build_report.rb]: Ugh, I really hate this syntax. It's not straight forward to understand that `pull_request` and `build` are params of `run`. What do you think about using plain Ruby for `self.run` definition?

File: app/services/build_report.rb [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+  def run
+    commenter.comment_on_violations(priority_violations)
+    create_success_status
+    track_subscribed_build_completed
+  end
+
+  private
+
+  def commenter
+    Commenter.new(pull_request)
```
Comment (gylaz) [inline app/services/build_report.rb]: What is `pull_request` defined? This feels like too much magic.

File: app/services/build_report.rb [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+  end
+
+  def token
+    ENV["HOUND_GITHUB_TOKEN"]
```
Comment (gylaz) [inline app/services/build_report.rb]: `fetch`?

File: app/services/build_runner.rb [hunk 12]
```diff
@@ -26,10 +24,8 @@ def review_pull_request
     track_subscribed_build_started
     create_pending_status
     upsert_owner
-    create_build
-    commenter.comment_on_violations(priority_violations)
-    create_success_status
-    track_subscribed_build_completed
+    build = create_build
+    BuildReport.run(pull_request, build)
```
Comment (gylaz) [inline app/services/build_runner.rb]: Thoughts on calling `create_build` inline?

File: app/services/build_report.rb (hunk 10 above)
Comment (teoljungberg) [inline app/services/build_report.rb]: It get's passed into the initializer from `static_facade`

File: app/services/build_report.rb (hunk 9 above)
Comment (teoljungberg) [inline app/services/build_report.rb]: I actually prefer it, but since we don't have it incorporated in the project I'm fine leaving it out

File: app/services/build_report.rb (hunk 11 above)
Comment (teoljungberg) [inline app/services/build_report.rb]: Good idea

File: spec/services/build_report_spec.rb [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+        build = create(:build, violations: build_list(:violation, 2))
+        commenter = stubbed_commenter
+        stubbed_github_api
+        pull_request = stubbed_pull_request
+
+        BuildReport.run(pull_request, build)
+
+        expect(commenter).to have_received(:comment_on_violations).
+          with(build.violations.take(BuildReport::MAX_COMMENTS))
+      end
```
Comment (gylaz) [inline spec/services/build_report_spec.rb]: Does this test duplicate what test on line 6 does? Can we remove that test?

File: spec/services/build_report_spec.rb [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+
+        BuildReport.run(pull_request, build)
+
+        expect(commenter).to have_received(:comment_on_violations).
+          with(build.violations)
+      end
+
+      it "comments a maximum number of times" do
+        stub_const("::BuildReport::MAX_COMMENTS", 1)
+        commenter = double("Commenter", comment_on_violations: true)
```
Comment (gylaz) [inline spec/services/build_report_spec.rb]: What do you think about modifying the `stubbed_commenter` function to take the changing variables? Similar to how we do that here: https://github.com/thoughtbot/hound/blob/master/spec/models/pull_request_spec.rb#L101-L108

File: spec/services/build_report_spec.rb [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+        allow(Commenter).to receive(:new).and_return(commenter)
+        build = create(:build, violations: build_list(:violation, 2))
+        commenter = stubbed_commenter
```
Comment (gylaz) [inline spec/services/build_report_spec.rb]: This doesn't look right.

File: spec/services/build_report_spec.rb [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+      it "creates GitHub statuses" do
+        repo = create(:repo, full_github_name: "test/repo")
+        build = create(
+          :build,
+          commit_sha: "headsha",
+          repo: repo,
+          violations: [
+            build(:violation),
+            build(:violation, messages: ["wrong", "bad"]),
+          ],
```
Comment (gylaz) [inline spec/services/build_report_spec.rb]: Is one violations enough to test that the status is created? Do we need any violations?

File: spec/services/build_runner_spec.rb [hunk 17]
```diff
@@ -29,34 +29,14 @@
...
       stubbed_github_api
+      pull_request = stubbed_pull_request
+      stubbed_style_checker_with_violations
+      allow(BuildReport).to receive(:run)
 
       build_runner.run
 
-      expect(commenter).to have_received(:comment_on_violations).
-        with(violations.take(BuildRunner::MAX_COMMENTS))
+      expect(BuildReport).to have_received(:run).with(pull_request, Build.last)
```
Comment (gylaz) [inline spec/services/build_runner_spec.rb]: How about updating the doc string of this test?

File: app/services/build_report.rb (hunk 10 above)
Comment (gylaz) [inline app/services/build_report.rb]: So, `static_facade` does two things? Define a `self.run` method, as well as create private `attr_reader`s?

File: spec/services/build_report_spec.rb (hunk 14 above)
Comment (teoljungberg) [inline spec/services/build_report_spec.rb]: :+1:

File: spec/services/build_report_spec.rb (hunk 15 above)
Comment (teoljungberg) [inline spec/services/build_report_spec.rb]: How so?

File: spec/services/build_report_spec.rb (hunk 16 above)
Comment (teoljungberg) [inline spec/services/build_report_spec.rb]: Good catch

File: app/services/build_report.rb (hunk 10 above)
Comment (teoljungberg) [inline app/services/build_report.rb]: That's correct

File: app/services/build_runner.rb (hunk 12 above)
Comment (teoljungberg) [inline app/services/build_runner.rb]: I would keep this as is - I don't think it's going to be clearer of inlining `create_build`

File: spec/services/build_report_spec.rb (hunk 15 above)
Comment (gylaz) [inline spec/services/build_report_spec.rb]: There's a `commenter` being created on line 20.

**Comment by teoljungberg** [comment]
@gylaz all comments are fixed and pushed. Ready for a last pass

File: spec/services/build_runner_spec.rb (hunk 17 above)
Comment (teoljungberg) [inline spec/services/build_runner_spec.rb]: Done in 9c52af4

File: spec/services/build_report_spec.rb (hunk 13 above)
Comment (teoljungberg) [inline spec/services/build_report_spec.rb]: Done in 7a2861e

File: spec/services/build_report_spec.rb [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+
+    context "with subscribed private repo and opened pull request" do
+      it "tracks build events" do
+        repo = create(:repo, :active, github_id: 123, private: true)
+        build = create(:build, repo: repo)
+        create(:subscription, repo: repo)
+        stubbed_commenter
+        stubbed_pull_request
+        stubbed_github_api
+        pull_request = stubbed_pull_request
```
Comment (gylaz) [inline spec/services/build_report_spec.rb]: Line 49 duplicates this.

**Comment by gylaz** [comment]
One more note, then good to merge.

**Comment by teoljungberg** [comment]
Thanks for reviewing @gylaz
